Modules/_functoolsmodule.c

```diff
@@ -120,11 +204,68 @@ partial_new(PyTypeObject *type, PyObject *args, PyObject *kw)
         Py_DECREF(pto);
         return NULL;
     }
-    if (pargs == NULL) {
+
+    pto->placeholder = PyObject_CallNoArgs((PyObject *)state->placeholder_type);
```

Comment:
Do you really need to store the placeholder object inside a partial object? it's a global module constant so I think you should be able to access it, or can't you?

Alternatively, to help in other functions, you could save the indices of the positional argumetns that are placeholders in the partialobject structure.

Reply:
Regarding the second point, I have thought about it and tried it. The thing is, it only saves few nanoseconds for function calls of medium size (e.g. 6-8 positional arguments), but it adds a considerable amount of complexity and extra data to be stored. Current C loop that handles putting things into position in `__call__` adds very minimal overhead. At least in C version.

Pre-storing indices does make Python version considerably faster and code much simpler. I.e. `operator.itemgetter` can be used efficiently. But Python version is of secondary importance here.

Will think about the first point. At the time when I did it, it made sense, but that was before `singleton` approach. Maybe just getting it from the module at run-time is better and will be more in line with Python version.

Reply:
I have tried different approaches to this, but can't see any other simple enough path here.

It is not possible to have it "global, global" due to the nature of Multi-phase module initialization. It has to be in module state.

So I inevitable have to retrieve it from there.

In initializer, it is not a big deal as it is done only once and also state can be conveniently retrieved from type.

From call, it would be one extra step and it would have to be done on every call.

Pre-storing indices is sensible strategy, but if python objects were used, it would slow things down and if to use a stack, then it would add a fair bit of code.

Essentially I agree with you that it would be better, but I don't see any good solution of which benefits would outweigh the effort and complexity added.

Currently it works well in all aspects, except the necessity to store it in object's struct.

Reply:
It would be good to have a dedicated module for constants, things would be a bit more straight forward in this case.

Reply:
BTW, this is python's solution that pre-stores indices:

```python
Placeholder = object()

def ppartial(func, *p_args, **p_kwds):
    np = p_args.count(Placeholder)
    if np:
        j = len(p_args)
        order = list(range(j))
        for i, a in enumerate(p_args):
            if a is Placeholder:
                order[i] = j
                j += 1
        ig = opr.itemgetter(*order)

    def wrapper(*args, **kwds):
        if np:
            n = len(args)
            if n < np:
                raise ValueError('Not all Placeholders are filled')
            f_args = ig(p_args + args[:np])
            args = args[np:] if n > np else ()
        else:
            f_args = p_args
        if p_kwds:
            kwds = {**p_kwds, **kwds}
        return func(*f_args, *args, **kwds)
    return wrapper
```

Reply:
When there is a place dedicated to similar constants it is easy to implement an independent constant and then remove it from `struct`. Until then, I don't see a good alternative to keeping `placeholder` in a `struct`. I am open to ideas, but resolving this for now.